File: python/src/spaturzu/_context.py
"""Frame propagation for per-agent attribution.

Mirrors `sdks/typescript/src/context.ts`: each call to ``spaturzu.run("name")``
opens a fresh frame, pushes it onto a ``contextvars.ContextVar`` (the Python
equivalent of Node's ``AsyncLocalStorage``), and pops it on exit. Frames
inherit ``run_id`` and extend ``agent_path``; child frame tags merge on top
of the parent's, so an inner ``run("nested", tags={"team": "ml"})`` overrides
the outer ``team`` value while keeping the rest.

Why a ContextVar (not threading.local): an asyncio Task gets its own context
copy, so concurrent tasks see independent frames after ``.set()`` in one
task. ``threading.local`` would only isolate threads, not coroutines.
"""

from __future__ import annotations

from contextvars import ContextVar, Token
from dataclasses import dataclass, field
from typing import Optional
from uuid import uuid4


@dataclass
class RunFrame:
    """Per-agent attribution state propagated through ``ContextVar``.

    ``last_request_id`` is intentionally mutable: each wrapped call sets it
    on the *current* frame so that a subsequent ``run()`` nested below
    captures it as the new frame's ``parent_request_id``. Mutation is safe
    because the frame is private to each ``ContextVar.set`` scope; sibling
    runs get distinct frames.
    """

    run_id: str
    agent_name: str
    agent_path: list[str]
    parent_request_id: Optional[str] = None
    last_request_id: Optional[str] = None
    tags: Optional[dict[str, str]] = None


# Sentinel default: frames are explicit, not inherited from process state.
_frame_var: ContextVar[Optional[RunFrame]] = ContextVar(
    "spaturzu_frame", default=None
)


def get_current_frame() -> Optional[RunFrame]:
    """Return the active frame, or ``None`` outside any ``spaturzu.run`` block."""
    return _frame_var.get()


def set_last_request_id(request_id: str) -> None:
    """Mark the most-recent request id on the current frame.

    Wrappers call this *before* the upstream HTTP call so that a concurrent
    nested ``run()`` reads it as ``parent_request_id``. No-op outside a frame.
    """
    frame = _frame_var.get()
    if frame is not None:
        frame.last_request_id = request_id


def merge_tags(
    a: Optional[dict[str, str]], b: Optional[dict[str, str]]
) -> Optional[dict[str, str]]:
    """Shallow-merge two tag bags; ``b`` wins on key conflict.

    Returns ``None`` if both are empty so callers can drop the field
    entirely instead of serialising ``{}``.
    """
    if not a and not b:
        return None
    if not a:
        return dict(b) if b else None
    if not b:
        return dict(a)
    out = dict(a)
    out.update(b)
    return out


def normalise_tags(
    tags: Optional[dict] = None,
) -> Optional[dict[str, str]]:
    """Coerce primitive tag values to strings; drop None/empty.

    Mirrors the Node `normalizeTags`. Returns None when nothing remains so
    callers can omit the field rather than serialise an empty dict.
    """
    if not tags:
        return None
    out: dict[str, str] = {}
    for k, v in tags.items():
        if v is None:
            continue
        out[k] = v if isinstance(v, str) else str(v)
    return out or None


@dataclass
class AgentBinding:
    """Binds a wrapped client to an agent name (+ optional frame-scoped tags)
    so each terminal call runs against a derived child frame. Produced by the
    `.with_agent()` accessor on a wrapped client."""

    agent: str
    tags: Optional[dict[str, str]] = None


def derive_frame(
    parent: Optional[RunFrame],
    agent_name: str,
    tags: Optional[dict[str, str]],
) -> RunFrame:
    """Build a child RunFrame from `parent` (run_id inherited or new,
    agent_path extended, parent_request_id from the parent's last_request_id,
    tags merged). Shared by `open_frame` and the `.with_agent()` capture path."""
    merged = merge_tags(parent.tags if parent else None, tags)
    return RunFrame(
        run_id=parent.run_id if parent else str(uuid4()),
        agent_name=agent_name,
        agent_path=([*parent.agent_path, agent_name] if parent else [agent_name]),
        parent_request_id=parent.last_request_id if parent else None,
        last_request_id=None,
        tags=merged,
    )


def open_frame(
    agent_name: str, tags: Optional[dict[str, str]] = None
) -> tuple[RunFrame, Token]:
    """Push a new frame onto the ContextVar.

    Returns ``(frame, token)``; the token must be passed to
    ``close_frame`` to restore the previous frame. Caller is responsible
    for pairing — the public API wraps this in a context manager so user
    code never has to.
    """
    parent = _frame_var.get()
    frame = derive_frame(parent, agent_name, tags)
    token = _frame_var.set(frame)
    return frame, token


def close_frame(token: Token) -> None:
    """Restore the parent frame using the token from ``open_frame``."""
    _frame_var.reset(token)

File: typescript/src/context.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { randomUUID } from "node:crypto";

/** Tags arrive from customers as primitives (env="prod", retries=3,
 *  is_internal=true). Coerce to string at the boundary so the wire format
 *  stays Record<string,string> and aggregations group cleanly. */
export type TagInput = Record<string, string | number | boolean>;

export function normalizeTags(
  t: TagInput | undefined,
): Record<string, string> | undefined {
  if (!t) return undefined;
  const out: Record<string, string> = {};
  let any = false;
  for (const [k, v] of Object.entries(t)) {
    if (v == null) continue;
    out[k] = typeof v === "string" ? v : String(v);
    any = true;
  }
  return any ? out : undefined;
}

/** Binds a wrapped client to an agent name (+ optional frame-scoped tags)
 *  so each terminal call runs inside `runInFrame(agent, ...)`. Will be
 *  produced by the `.withAgent()` accessor on a wrapped client (added later
 *  in this feature; see the wrap modules). */
export interface AgentBinding {
  agent: string;
  tags?: Record<string, string>;
}

/** Return type of the `wrapX` functions: the wrapped client `T` plus a
 *  `.withAgent()` accessor that tags each terminal call with an agent name
 *  (and optional frame-scoped tags). Chainable — `.withAgent()` returns
 *  another `WithAgent<T>`. Provider-agnostic, so it lives here rather than
 *  being duplicated per wrap module. */
export type WithAgent<T> = T & {
  withAgent(agent: string, opts?: { tags?: TagInput }): WithAgent<T>;
};

// A RunFrame is the unit of agent attribution that propagates through
// AsyncLocalStorage. It survives `await`, `Promise.all`, `setTimeout`, and
// every other async boundary Node tracks.
//
// `lastRequestId` is intentionally mutable: each wrapped LLM call sets it on
// the *current* frame so that a subsequent `spaturzu.run(...)` nested below
// picks it up as the new frame's `parentRequestId`. Mutation is safe because
// the frame object is private to each call to `als.run(frame, fn)` — sibling
// runs get distinct frames.
export interface RunFrame {
  /** Stable across an entire workflow execution. */
  runId: string;
  /** Logical agent name for the current frame. */
  agentName: string;
  /** Full chain from root to current frame, e.g. ["research","writer"]. */
  agentPath: string[];
  /** ID of the parent agent's most recent request at the moment this frame
   *  was opened. Stays constant for every call made by this agent — points
   *  to "the call that spawned me." */
  parentRequestId?: string;
  /** ID of the most recent LLM call started inside this frame. Mutated by
   *  the wrappers; consumed by `run()` when nesting a child agent. */
  lastRequestId?: string;
  /** Cumulative frame-scoped tags. Inherits from the parent frame (if any)
   *  and extends with whatever was passed to `run()`. The wrappers further
   *  merge in process-wide spaturzu constructor tags at log-build time —
   *  this is just the run-scoped layer. */
  tags?: Record<string, string>;
}

const storage = new AsyncLocalStorage<RunFrame>();

export function getCurrentFrame(): RunFrame | undefined {
  return storage.getStore();
}

/** Run `fn` inside a fresh agent frame. If called inside another frame,
 *  the new frame inherits the runId and extends the agent path; the parent's
 *  lastRequestId becomes the child's parentRequestId.
 *
 *  Tags: the new frame inherits the parent frame's tags and extends them
 *  with `tags` (passed-in keys win on conflict — the inner scope is the
 *  more specific one). Pass `undefined` to inherit cleanly. */
export function runInFrame<T>(
  agentName: string,
  fn: () => Promise<T>,
  tags?: Record<string, string>,
): Promise<T> {
  const parent = getCurrentFrame();
  const mergedTags = mergeTags(parent?.tags, tags);
  const frame: RunFrame = {
    runId: parent?.runId ?? randomUUID(),
    agentName,
    agentPath: parent
      ? [...parent.agentPath, agentName]
      : [agentName],
    parentRequestId: parent?.lastRequestId,
    lastRequestId: undefined,
    tags: mergedTags,
  };
  return storage.run(frame, fn);
}

/** Shallow-merge two tag bags, with `b` winning on key conflict. Returns
 *  `undefined` if both are empty so callers can drop the field entirely
 *  rather than serialising `{}`. */
export function mergeTags(
  a: Record<string, string> | undefined,
  b: Record<string, string> | undefined,
): Record<string, string> | undefined {
  if (!a && !b) return undefined;
  if (!a) return b ? { ...b } : undefined;
  if (!b) return { ...a };
  return { ...a, ...b };
}

/** Set the most-recent-call pointer on the current frame, if any. Called
 *  by wrappers as soon as they generate a request UUID, *before* the
 *  upstream HTTP call — so concurrent nested runs see it immediately. */
export function setLastRequestId(requestId: string): void {
  const frame = getCurrentFrame();
  if (frame) frame.lastRequestId = requestId;
}
